[rpc] remove default constructor in futureInfo

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30197 [rpc] remove default constructor in futureInfo**

Addresses @pietern's comment on https://github.com/pytorch/pytorch/pull/29601. This default constructor was added because `std::map`'s `operator[]` requires a default constructor. However, instead of using `operator[]`, we can
use `emplace` and remove the constructor, to ensure that the `FutureInfo` struct
doesn't get constructed with garbage values.

Differential Revision: [D18627675](https://our.internmc.facebook.com/intern/diff/D18627675/)

rohan-varma committed:
[rpc] remove default constructor in futureInfo

This default constructor was added because std::map's operator[]
requires a default constructor. However, instead of using operator[], we can
use emplace and remove the constructor, to ensure that the FutureInfo struct
doesnt get constructed with garbage values.

Differential Revision: [D18627675](https://our.internmc.facebook.com/intern/diff/D18627675/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -93,7 +93,6 @@ class ProcessGroupAgent : public RpcAgent {
           startTime_(startTime),
           dstRank_(dstRank),
           timeout_(timeout) {}
-    FutureInfo() {}
```
Delete explicitly?
```cpp
FutureInfo() = delete;
```

xush6528 replied:
```cpp
FutureInfo() = delete;
```

zhaojuanmao commented:
@rohan-varma  I just caught up diff review about future time out PR in #29601, I think we do not need startTime_ and timeout_, endTime_ should be sufficient.

pietern commented:
Thanks, @rohan-varma. This makes it just a bit more robust for future changes.

xush6528 commented:
@zhaojuanmao 

I am doing it in https://github.com/pytorch/pytorch/pull/30342.

rohan-varma committed:
Update on "[rpc] remove default constructor in futureInfo"


Addresses @pietern's comment on https://github.com/pytorch/pytorch/pull/29601. This default constructor was added because `std::map`'s `operator[]` requires a default constructor. However, instead of using `operator[]`, we can
use `emplace` and remove the constructor, to ensure that the `FutureInfo` struct
doesn't get constructed with garbage values.

Differential Revision: [D18627675](https://our.internmc.facebook.com/intern/diff/D18627675/)

[ghstack-poisoned]

pietern reviewed (approved):
LGTM, thank you!
